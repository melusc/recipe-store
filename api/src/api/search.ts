/*!
	This program is free software: you can redistribute it
	and/or modify it under the terms of the GNU General Public
	License as published by the Free Software Foundation,
	either version 3 of the License, or (at your option)
	any later version.

	This program is distributed in the hope that it will be
	useful, but WITHOUT ANY WARRANTY; without even the implied
	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
	PURPOSE. See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public
	License along with this program. If not, see <https://www.gnu.org/licenses/>.
*/

import {sectionToText} from 'cooklang';

import type {Recipe} from './recipe.js';

type Qualifier = 'any' | 'tagged' | 'author' | 'contains' | 'title';

export type QueryFilter = {
	readonly qualifier: Qualifier;
	readonly filterValue: string;
};

const qualifiers: Record<string, Qualifier> = {
	tagged: 'tagged',
	tag: 'tagged',
	title: 'title',
	titled: 'title',
	intitle: 'title',
	author: 'author',
	user: 'author',
	by: 'author',
	contains: 'contains',
	intext: 'contains',
};

/*
 * - "Qualifier" is the type of filter `tagged`, `author`, etc
 * - No backslashes or so for escaping quotes
 * - Unmatched quotes are okay.
 *   It just means it will treat the rest of the string as part of that value
 * - A filter can be quoted "tagged:abc" or just the filter-value tagged:"abc"
 * - qualifier and filters can be seperated by whitespace
 */

export class QueryParser {
	index = 0;

	constructor(public readonly query: string) {}

	hasChar() {
		return this.index < this.query.length;
	}

	peek() {
		return this.query[this.index] ?? '';
	}

	readChar() {
		if (this.index >= this.query.length) {
			throw new TypeError('Tried to read character out of bounds.');
		}

		const character = this.query[this.index]!;
		this.next();
		return character;
	}

	next() {
		++this.index;
	}

	skipWS() {
		while (this.hasChar() && /\s/.test(this.peek())) {
			this.next();
		}
	}

	readUntil(stopCondition: string | RegExp) {
		const shouldStop =
			typeof stopCondition === 'string'
				? (char: string) => char === stopCondition
				: (char: string) => stopCondition.test(char);

		let result = '';
		while (this.hasChar() && !shouldStop(this.peek())) {
			result += this.readChar();
		}

		return result;
	}

	readQualifierFilter() {
		this.skipWS();

		if (this.peek() === '"' || this.peek() === "'") {
			const stop = this.readChar();
			const result = this.readUntil(stop);
			this.next();
			return result;
		}

		return this.readUntil(/[\s"']/);
	}

	readQuotedValue(): QueryFilter {
		const stop = this.readChar();
		const filter = this.readUntil(stop);
		this.next();

		if (filter.includes(':')) {
			const [left, ...search] = filter.split(':');
			const qualifier = qualifiers[left!.toLowerCase()] ?? 'any';
			return {
				qualifier,
				filterValue: search.join(':'),
			};
		}
		return {
			qualifier: 'any',
			filterValue: filter,
		};
	}

	parse() {
		this.index = 0;

		const filters: QueryFilter[] = [];

		while ((this.skipWS(), this.hasChar())) {
			if (this.peek() === '"' || this.peek() === "'") {
				filters.push(this.readQuotedValue());

				continue;
			}

			const part = this.readUntil(/[:"'\s]/);

			if (this.peek() === ':') {
				this.next();
				const filter = this.readQualifierFilter();

				const qualifier = qualifiers[part.toLowerCase()] ?? 'any';
				filters.push({
					qualifier,
					filterValue: filter,
				});
			} else {
				filters.push({
					qualifier: 'any',
					filterValue: part,
				});
			}
		}

		return filters;
	}
}

function normaliseSearchValue(value: string) {
	// Treat all whitespace equal by normalising it to a space
	// "abc\r\ndef" == "abc\tdef"
	// Ignore punctuation: "Firstly, add" == "firstly add"
	// Don't care about case
	return value.toLowerCase().replaceAll(/[\s,.:;\-!?]+/g, ' ');
}

function searchContains(needle: string, haystack: string) {
	needle = normaliseSearchValue(needle);
	haystack = normaliseSearchValue(haystack);

	return haystack.includes(needle);
}

const filterMatchers = {
	any(this, filterValue: string, recipe: InstanceType<Recipe>) {
		// `contains:` last because it is probably the slowest, because it has to stringify the sections
		return (
			this.title(filterValue, recipe) ||
			this.tagged(filterValue, recipe) ||
			this.author(filterValue, recipe) ||
			this.contains(filterValue, recipe)
		);
	},
	title(filterValue: string, recipe: InstanceType<Recipe>) {
		return searchContains(filterValue, recipe.title);
	},
	tagged(filterValue: string, recipe: InstanceType<Recipe>) {
		return recipe.tags.some(tag => searchContains(filterValue, tag));
	},
	author(filterValue: string, recipe: InstanceType<Recipe>) {
		if (!recipe.author) {
			return false;
		}

		// author:32
		const asUserId = Number.parseInt(filterValue);
		if (asUserId === recipe.author.userId) {
			return true;
		}

		return (
			// author:Michael Caine
			searchContains(filterValue, recipe.author.displayName) ||
			// author:michaelcaine33
			searchContains(filterValue, recipe.author.username)
		);
	},
	contains(filterValue: string, recipe: InstanceType<Recipe>) {
		// contains:"add to bowl"
		for (const section of recipe.sections) {
			const stringified = sectionToText(section.parsed);
			if (searchContains(filterValue, stringified)) {
				return true;
			}
		}

		return false;
	},
} as const;

export function recipeMatchesFilter(
	recipe: InstanceType<Recipe>,
	filters: readonly QueryFilter[],
) {
	return filters.every(filter =>
		filterMatchers[filter.qualifier](filter.filterValue, recipe),
	);
}
