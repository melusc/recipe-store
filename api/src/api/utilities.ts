/*!
	This program is free software: you can redistribute it
	and/or modify it under the terms of the GNU General Public
	License as published by the Free Software Foundation,
	either version 3 of the License, or (at your option)
	any later version.

	This program is distributed in the hope that it will be
	useful, but WITHOUT ANY WARRANTY; without even the implied
	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
	PURPOSE. See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public
	License along with this program. If not, see <https://www.gnu.org/licenses/>.
*/

export type ReadonlyDate = Readonly<Omit<Date, `set${string}`>>;

export class PaginationResult<T> {
	public readonly pageCount: number;
	public readonly page: number;
	public readonly items: readonly T[];

	constructor({
		pageCount,
		page,
		items,
	}: {
		readonly pageCount: number;
		readonly page: number;
		readonly items: readonly T[];
	}) {
		this.pageCount = pageCount;
		this.page = page;
		this.items = items;
	}

	getNextPage() {
		if (this.page >= this.pageCount) {
			return false;
		}

		// If page is negative for some reason,
		// skip to page 1
		return Math.max(1, this.page + 1);
	}

	getPreviousPage() {
		if (this.page <= 1) {
			return false;
		}

		// If page is way too high
		// skip to last page
		return Math.min(this.pageCount, this.page - 1);
	}
}
