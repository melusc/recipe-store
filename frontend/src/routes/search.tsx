/*!
	Copyright 2025 Luca Schnellmann <oss@lusc.ch>

	This file is part of recipe-store.

	This program is free software: you can redistribute it
	and/or modify it under the terms of the GNU General Public
	License as published by the Free Software Foundation,
	either version 3 of the License, or (at your option)
	any later version.

	This program is distributed in the hope that it will be
	useful, but WITHOUT ANY WARRANTY; without even the implied
	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
	PURPOSE. See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public
	License along with this program. If not, see <https://www.gnu.org/licenses/>.
*/

import {RelativeUrl} from '@lusc/util/relative-url';
import type {DynamicPaginationResult, Recipe} from 'api';

import {Pagination} from '../components/pagination.js';
import {RecipeCard} from '../components/recipe-card.js';
import {SearchForm} from '../components/search-form.js';

import {createRoute} from './_utilities.js';

export const renderSearch = createRoute<
	| [query: undefined, recipes: undefined]
	| [query: string, recipes: DynamicPaginationResult<Recipe>]
>((_, query, recipes) => {
	if (query === undefined) {
		return {
			title: 'Search',
			body: (
				<main>
					<SearchForm location="inline" />
				</main>
			),
		};
	}

	const paginationBaseUrl = new RelativeUrl('/search');
	paginationBaseUrl.searchParams.set('q', query);
	const paginationButtons = (
		<Pagination baseUrl={paginationBaseUrl} paginationResult={recipes!} />
	);

	if (recipes!.items.length === 0) {
		return {
			title: query,
			body: (
				<>
					<main>
						<SearchForm location="inline" prefillQuery={query} />
					</main>

					{paginationButtons}
				</>
			),
		};
	}

	return {
		title: query,
		body: (
			<>
				<main class="row g-3">
					<SearchForm location="inline" prefillQuery={query} />

					{recipes!.items.map(recipe => (
						<RecipeCard recipe={recipe} />
					))}
				</main>

				{paginationButtons}
			</>
		),
	};
});
