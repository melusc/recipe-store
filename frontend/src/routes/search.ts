/*!
	This program is free software: you can redistribute it
	and/or modify it under the terms of the GNU General Public
	License as published by the Free Software Foundation,
	either version 3 of the License, or (at your option)
	any later version.

	This program is distributed in the hope that it will be
	useful, but WITHOUT ANY WARRANTY; without even the implied
	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
	PURPOSE. See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public
	License along with this program. If not, see <https://www.gnu.org/licenses/>.
*/

import {RelativeUrl} from '@lusc/util/relative-url';
import {type DynamicPaginationResult, type Recipe} from 'api';

import {$} from '../$.js';
import {pagination} from '../components/pagination.js';
import {recipeCard} from '../components/recipe-card.js';
import {searchForm} from '../components/search-form.js';

import {createRoute} from './_utilities.js';

export const renderSearch = createRoute<
	| [query: undefined, recipes: undefined]
	| [query: string, recipes: DynamicPaginationResult<Recipe>]
>('Search', (_user, query, recipes) => {
	if (query === undefined) {
		return $`
			<main>
				${searchForm('inline')}
			</main>
		`;
	}

	const paginationBaseUrl = new RelativeUrl('/search');
	paginationBaseUrl.searchParams.set('q', query);
	const paginationButtons = pagination(paginationBaseUrl, recipes!);

	if (recipes!.items.length === 0) {
		return $`
			<main>
				${searchForm('inline', query)}
			</main>

			${paginationButtons}
		`;
	}

	return $`
		<main class="row g-3">
			${searchForm('inline', query)}

			${recipes!.items.map(recipe => recipeCard(recipe))}
		</main>

		${paginationButtons}
	`;
});
