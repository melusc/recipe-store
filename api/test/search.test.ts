/*!
	This program is free software: you can redistribute it
	and/or modify it under the terms of the GNU General Public
	License as published by the Free Software Foundation,
	either version 3 of the License, or (at your option)
	any later version.

	This program is distributed in the hope that it will be
	useful, but WITHOUT ANY WARRANTY; without even the implied
	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
	PURPOSE. See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public
	License along with this program. If not, see <https://www.gnu.org/licenses/>.
*/

import {describe, expect, test} from 'vitest';

import {QueryParser, recipeMatchesFilter} from '../src/api/search.js';
import {UserRoles} from '../src/index.js';

import {apiTest} from './utilities.js';

describe('QueryParser', () => {
	test.for<string>([
		'abc',
		'tag:abc',
		'"contains:abc"',
		`'by:abc"'`,
		'author:"abc def"',
		'abc def',
		"'abc def'",
		'"abc def"',
		'"abc',
		'   intext:abc  ',
		'   user: abc ',
		'tagged:contains:ABC',
		'tagged:cookie author:Sophia',
		'tagged: "abc"',
		'tagged:abc"def"',
		'title:coolyo',
		'intitle:pie',
		'TitLeD: "soft cake"',
		'',
		' ',
	])('parse(%j)', query => {
		const queryParser = new QueryParser(query);
		const parsed = queryParser.parse();

		expect(parsed).toMatchSnapshot();
	});
});

apiTest.for<[query: string, ...expectedRecipeTitles: string[]]>([
	['user:1', 'Banana Cake', 'Mushroom Risotto'],
	['user:1 waiter', 'Mushroom Risotto'],
	['user:2', 'Chocolate Cake'],
	['user:Michael', 'Banana Cake', 'Mushroom Risotto'],
	['user:pickle', 'Chocolate Cake'],
	['tagged:cake', 'Banana Cake', 'Chocolate Cake'],
	['tagged:"chocolate cake"', 'Chocolate Cake'],
	['title:mushroom', 'Mushroom Risotto'],
	['easy', 'Banana Cake'],
	['restaurant', 'Mushroom Risotto'],
	['first prepare', 'Banana Cake'],
	['unique', 'Banana Cake', 'Chocolate Cake'],
	['tagged:unique', 'Banana Cake'],
	['contains:unique', 'Chocolate Cake'],
	['unique banana', 'Banana Cake'],
])(
	'recipeMatchesFilter(%j)',
	async ([query, ...expectedTitles], {api: {User, Recipe}}) => {
		const expectedTitlesSet = new Set(expectedTitles);
		const queryParsed = new QueryParser(query).parse();

		const user1 = User.create(
			'michaelcaine33',
			'Michael Caine',
			'vvnem',
			UserRoles.User,
		);
		const user2 = User.create(
			'picklerick',
			'Rick Sanchez',
			'emaee',
			UserRoles.User,
		);

		const recipes = await Promise.all([
			Recipe.create(
				'Banana Cake',
				user1,
				undefined,
				// "unique" appears only in tags
				['cake', 'banana', 'dessert', 'easy cakes', 'unique'],
				[
					'First, prepare a bowl with @flour{200%g} and @sugar{150%g}. Stir until uniform.',
					'Add @bananas{3} to a seperate bowl and lightly mash.',
					'Combine the two bowls and add @milk{200%mL}.',
				],
			),
			Recipe.create(
				'Chocolate Cake',
				user2,
				undefined,
				// Badly tagged
				['chocolate cake'],
				[
					'Buy @chocolate cake{} from shop.',
					'Put in #fridge{} and let cool down for a few hours.',
					// "unique" appears only in text
					'unique',
				],
			),
			Recipe.create(
				'Mushroom Risotto',
				user1,
				undefined,
				['risotto', 'mushroom'],
				['Go to #restaurant{}.', 'Ask waiter for risotto with mushroom.'],
			),
		]);

		const matchedRecipes = recipes.filter(recipe =>
			recipeMatchesFilter(recipe, queryParsed),
		);
		const matchedRecipesTitles = new Set(
			matchedRecipes.map(recipe => recipe.title),
		);

		expect(matchedRecipesTitles).toStrictEqual(expectedTitlesSet);
	},
);
