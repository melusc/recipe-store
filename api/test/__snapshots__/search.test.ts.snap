// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QueryParser > parse("   intext:abc  ") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "contains",
  },
]
`;

exports[`QueryParser > parse("   user: abc ") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "author",
  },
]
`;

exports[`QueryParser > parse(" ") 1`] = `[]`;

exports[`QueryParser > parse("") 1`] = `[]`;

exports[`QueryParser > parse("'abc def'") 1`] = `
[
  {
    "filterValue": "abc def",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("'by:abc\\"'") 1`] = `
[
  {
    "filterValue": "abc"",
    "invert": false,
    "qualifier": "author",
  },
]
`;

exports[`QueryParser > parse("\\"-abc\\"") 1`] = `
[
  {
    "filterValue": "-abc",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("\\"abc def\\"") 1`] = `
[
  {
    "filterValue": "abc def",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("\\"abc") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("\\"contains:abc\\"") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "contains",
  },
]
`;

exports[`QueryParser > parse("- abc") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": true,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("-\\"abc\\"") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": true,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("-abc -def") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": true,
    "qualifier": "any",
  },
  {
    "filterValue": "def",
    "invert": true,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("-abc def") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": true,
    "qualifier": "any",
  },
  {
    "filterValue": "def",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("-abc") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": true,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("-tagged:abc") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": true,
    "qualifier": "tagged",
  },
]
`;

exports[`QueryParser > parse("TitLeD: \\"soft cake\\"") 1`] = `
[
  {
    "filterValue": "soft cake",
    "invert": false,
    "qualifier": "title",
  },
]
`;

exports[`QueryParser > parse("abc -\\"def\\"") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "any",
  },
  {
    "filterValue": "def",
    "invert": true,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("abc -def") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "any",
  },
  {
    "filterValue": "def",
    "invert": true,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("abc def") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "any",
  },
  {
    "filterValue": "def",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("abc") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("author:\\"abc def\\"") 1`] = `
[
  {
    "filterValue": "abc def",
    "invert": false,
    "qualifier": "author",
  },
]
`;

exports[`QueryParser > parse("intitle:pie") 1`] = `
[
  {
    "filterValue": "pie",
    "invert": false,
    "qualifier": "title",
  },
]
`;

exports[`QueryParser > parse("source:google.com") 1`] = `
[
  {
    "filterValue": "google.com",
    "invert": false,
    "qualifier": "source",
  },
]
`;

exports[`QueryParser > parse("source:https://google.com/") 1`] = `
[
  {
    "filterValue": "https://google.com/",
    "invert": false,
    "qualifier": "source",
  },
]
`;

exports[`QueryParser > parse("source:https://google.com/?q=rick%20roll") 1`] = `
[
  {
    "filterValue": "https://google.com/?q=rick%20roll",
    "invert": false,
    "qualifier": "source",
  },
]
`;

exports[`QueryParser > parse("tag:abc") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "tagged",
  },
]
`;

exports[`QueryParser > parse("tagged: \\"abc\\"") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "tagged",
  },
]
`;

exports[`QueryParser > parse("tagged:abc\\"def\\"") 1`] = `
[
  {
    "filterValue": "abc",
    "invert": false,
    "qualifier": "tagged",
  },
  {
    "filterValue": "def",
    "invert": false,
    "qualifier": "any",
  },
]
`;

exports[`QueryParser > parse("tagged:contains:ABC") 1`] = `
[
  {
    "filterValue": "contains:ABC",
    "invert": false,
    "qualifier": "tagged",
  },
]
`;

exports[`QueryParser > parse("tagged:cookie author:Sophia") 1`] = `
[
  {
    "filterValue": "cookie",
    "invert": false,
    "qualifier": "tagged",
  },
  {
    "filterValue": "Sophia",
    "invert": false,
    "qualifier": "author",
  },
]
`;

exports[`QueryParser > parse("title:coolyo") 1`] = `
[
  {
    "filterValue": "coolyo",
    "invert": false,
    "qualifier": "title",
  },
]
`;
